keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+its due diligence to find nodes to place all pending Pods. However, in a real-world case,
+some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
+actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
+unnecessary manner.
+
+Lacking the knob to flag Pods as schedule-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal comes up with an extensible mechanics to ensure Pods are scheduled only when they are
+semantically ready.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.status.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.
+
+For Pods carrying `.status.schedulePause=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only gets retried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an application developer, I'm knowledgeable to know when my Pods are schedule-ready, but I know
+nothing about scheduling internals. It'd be ideal for me to only set/unset the "paused" flag,
+instead of developing scheduler plugins, to control the schedulability of my Pods.
+
+#### Story 2
+
+As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
+to guard the schedulability of my Pods.
```

Comment:
I think we need to clarify why the pauseScheduling flag is not enough to address this story. One could argue that an external controller could be built to set the pauseScheduling to "guard the schedulability of my pods"

Reply:
Updated PTAL.

Reply:
For the motivation #2 and #3 in the discussion doc (https://docs.google.com/document/d/1cPxF3wIbGRw_inv4CCG6z6qC4WjHAPR5GLT__dRBx7I/edit#heading=h.whhghc8mxakf), how would a user guard schedulability of pods on independent conditions? An API that allows for some dynamic conditions (like an array of scheduling guards, similar to pod readiness gates design) would allow for this, rather than a single boolean.

> One could argue that an external controller could be built to set the pauseScheduling to "guard the schedulability of my pods"
I agree, a user can create a unified controller to control flipping the flag (eg: ensure that all dependent secrets are available, and dependent PVC is bound), but that makes the logic dependent, and prevents the logic from being distributed. Say the user wants to pause scheduling on two conditions, 1) PVCs are bound and 2) a custom condition related to networking configuration being available to the cluster. The first condition is fairly generic and can apply to any Stateful workload, the second is custom user logic.

Reply:
@pwschuurman are you supportive of the 2nd story or not? I guess yes :)

Just as you said, some advanced developers may choose to ignore the binary flag, and implement their own combinations of enqueue plugins, which is exactly the motivation of the 2nd story. In other words, in this story, you don't need the API field to achieve the goal. But this needs extra expertise on scheduler framework, so the first story and API field are still necessary.

Generally, story 1 and story 2 are complementary to each other to fit different personas and use cases.

Reply:
I see, so for the 2nd story, an advanced user could leverage the Scheduling Plugin API today (as long as their Kubernetes cluster supports the Plugin Scheduling Framework).

> Generally, story 1 and story 2 are complementary to each other to fit different personas and use cases.

And just to confirm, a user following the 1st story using this API field can complement existing Scheduling plugins set up by say a Kubernetes cluster administrator (eg: an existing plugin installed on the cluster as a default).

Reply:
That's correct. The general rules to how to cooperate with different enqueue plugins are documented in the Design#Implementations section, some teeny tiny details may need to be polished/reviewed in the future code PR though.